km/km_filesys.c

```diff
@@ -140,122 +171,188 @@ int km_active_accept(void)
-          0) {
-         km_err(2, "can't get listen socket fd protocol");
-      }
-      snap_listen_sockinfo->addrlen = sizeof(snap_listen_sockinfo->addr);
-      if (getsockname(snap_listen_sock,
-                      (struct sockaddr*)snap_listen_sockinfo->addr,
-                      (socklen_t*)&snap_listen_sockinfo->addrlen) < 0) {
-         km_err(2, "getsockname on listen socket fd");
-      }
-
-      // Get rid of vm related fd's and payload fd's we inherited from the previous instance of km
-      km_vmstate_destroy(snap_listen_sock, elf_fd);
-   } else {
-      km_errx(2, "env var %s's value %s is invalid", KM_SNAP_LISTEN_PORT, port);
+   // Build a list of the listening socket fd's in the snapshot
+   km_snapshot_listenfds_find(e);
+   km_infox(KM_TRACE_SHRINK,
+            "light_snap_accept_timeout %ld ms, listener count %d",
+            light_snap_accept_timeout,
+            km_snap_listening_state_cnt);
+   if (km_snap_listening_state_cnt <= 0) {
+      // No listening fd's in the snapshot.
+      // We have nothing to trigger shrinking an idle running snapshot so
+      // just run as normal.
+      km_infox(KM_TRACE_SNAPSHOT, "snapshot has no listening sockets to trigger payload shrink");
+      return;
    }
 
-   // will adjust the backlog value to that of the snapshot later
-   if (listen(snap_listen_sock, 1) < 0) {
-      km_err(2, "listen failed, fd %d", snap_listen_sock);
+   // Create open fd's for all the listening sockets.  If we got here because
+   // the running snapshot shrank itself, then the listening socket fd's will
+   // be open so we don't need to open anything.
+   // It is important that earlier parts of km do NOT use the payload fd space.
+   // That fd may be one of the fd's that the resuming snapshot will be listening
+   // on and the existence of the open fd will confuse this into thinking we are
+   // here because of a shrink operation.
+   for (int i = 0; i < km_snap_listening_state_cnt; i++) {
+      km_infox(KM_TRACE_SHRINK,
+               "%d: listen_fd %d, accept_fd %d, port %d",
+               i,
+               km_snap_listening_state_p[i].listen_fd,
+               km_snap_listening_state_p[i].accept_fd,
+               ntohs(((struct sockaddr_in*)km_snap_listening_state_p[i].listen_sockinfo.addr)->sin_port));
+      struct stat statb;
+      if (fstat(km_snap_listening_state_p[i].listen_fd, &statb) == 0) {
+         km_assert((statb.st_mode & S_IFMT) == S_IFSOCK);
+         // We've been exec'ed to by a payload shrink, the payload's fd's are still open.
+         // Get rid of vm related fd's and payload fd's we inherited from the previous instance of
+         // km But, we keep the payload's listening fd's and the open snapshot.
+         km_vmstate_destroy(fileno(e->file));
+         km_infox(KM_TRACE_SNAPSHOT,
+                  "Using %d listening fd's that existed before shrinking",
+                  km_snap_listening_state_cnt);
+         break;
```

Comment:
Are we assuming here that all listeners are open or all are closed? This is a bit confusing written like this.